R/teal_lockfile.R

```diff
@@ -39,50 +39,23 @@ teal_lockfile <- function() {
     }
   }
 
-  if (!(is_in_test() || is_r_cmd_check())) {
-    old_plan <- future::plan()
-    # If there is already a parallel (non-sequential) backend, reuse it.
-    if (inherits(old_plan, "sequential")) {
-      future::plan(future::multisession, workers = 2)
-    }
-
-    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
-    lockfile_task$invoke(close = inherits(old_plan, "sequential"), lockfile_path)
-    logger::log_trace("lockfile creation invoked.")
-  }
+  shiny::onStop(function() file.remove(lockfile_path))
+  callr::r_bg(create_renv_lockfile, args = list(lockfile_path = lockfile_path))
+  logger::log_trace("lockfile creation started.")
 }
 
-create_renv_lockfile <- function(close = FALSE, lockfile_path = NULL) {
-  checkmate::assert_flag(close)
+create_renv_lockfile <- function(lockfile_path = NULL) {
   checkmate::assert_string(lockfile_path, na.ok = TRUE)
-  promise <- promises::future_promise({
-    # Below we can not use a file created in tempdir() directory.
-    # If a file is created in tempdir(), it gets deleted on 'then(onFulfilled' part.
-    shiny::onStop(function() file.remove(lockfile_path))
 
-    renv_logs <- utils::capture.output(
-      renv::snapshot(
-        lockfile = lockfile_path,
-        prompt = FALSE,
-        force = TRUE
-        # type = is taken from renv::settings$snapshot.type()
-      )
+  renv_logs <- utils::capture.output(
+    renv::snapshot(
+      lockfile = lockfile_path,
+      prompt = FALSE,
+      force = TRUE
+      # type = is taken from renv::settings$snapshot.type()
     )
-    if (any(grepl("Lockfile written", renv_logs))) {
-      logger::log_trace("lockfile created successfully.")
-    } else {
-      logger::log_trace("lockfile created with issues.")
-    }
```

Comment:
How would you implement this on a feature branch? I don't see immediately  how this could be done.

Reply:
I actually don't think that's possible..
Even if `create_renv_lockfile` would return `renv` message with the status, something at the top level of `teal_lockfile` would need to wait until `create_renv_lockfile` is done to pass the message to the `logger`.

Reply:
Ok, fair enough. I think it is more important to not change `future::plan` then caring so much about this log

Reply:
@gogonzo I was able to bring back logging with this commit e04fff129ce0088b9d28748ca4d4d543957c798b
However there is a lot of information that you need to suppress if you want the `r_bg` function to print results to the console. It has it's own message and it's own custom log, so you need to run `capture.output` twice.

You can experiment with

```r
  x <- 
    utils::capture.output(
      utils::capture.output(
        r_bg(
          teal:::create_renv_lockfile, 
          args = list(lockfile_path = 'marcin.lock', opts = options()),
          stdout = ""
        ),
        type = "message"
      ),
    type = "output"
  )
```

The first `capture.output` is needed to suppress: `'Opening fd 1'` message
The second `capture.output` is needed to suppress: `'PROCESS 'Rterm', running, pid'` output

Reply:
Simplified it so we only run

```r
utils::capture.output(
        r_bg(
          teal:::create_renv_lockfile, 
          args = list(lockfile_path = 'marcin.lock', opts = options()),
          stdout = ""
        ),
        type = "message"
      )
```

to capture `'Opening fd 1'` message

Reply:
We stored the `callr` in a variable that is accessible in `init`. Last thing is to use `reactivePoll` or `reactiveTimer` to detect that the process finished, so that we can print logs about the finalization of the process.

Reply:
Useful internet resources to read and watch 
https://www.youtube.com/watch?v=DTMVzK7iZFU&t=1629s
https://towardsdev.com/a-simple-workflow-for-async-shiny-with-callr-f32393b087be